File: DynamicProgramming/Football.java
public class Football {
    public static int recursion(int score) {
        if (score == 0)
            return 1;
        if (score < 0)
            return 0;
        return recursion(score - 7) + recursion(score - 3) + recursion(score - 2);
    }

    public static int countCombinations(int score) {
        int comb[] = new int[score + 1];
        comb[0] = 1;
        for (int i = 1; i <= score; ++i) {
            int c = 0;
            if (i >= 7)
                c += comb[i - 7];
            if (i >= 3)
                c += comb[i - 3];
            if (i >= 2)
                c += comb[i - 2];
            comb[i] = c;
        }

        return comb[score];
    }

    public static void main(String[] args) {
        for (int i = 0; i < 20; ++i)
            System.out.println(i + " " + countCombinations(i) + " " + recursion(i));
    }
}

File: DynamicProgramming/football/football.c
/**********************************************************
 *
 *  Football
 *
 *********************************************************/
#include <stdio.h>
#include <stdlib.h>

int countCombinations(int score) {
    int *comb =  malloc(sizeof(int) * (score + 1));
    comb[1] = 0;
    comb[0] = comb[2] = comb[3] = comb[4] = 1;
    comb[5] = comb[6] =  2;
    comb[7] = 4;
    if (score <= 7)
        return comb[score];
    for(int i = 8; i <= score; ++i)
        comb[i] = comb[i - 7] + comb[i - 3] + comb[i - 2]; 
    
    int result = comb[score];
    free(comb);
    return result;
}

int main() {
    for(int i = 0; i < 20; ++i)
        printf("%d %d\n", i, countCombinations(i));

    return 0;
} 
